### PR TITLE
Add license header check for rust files

### PR DESCRIPTION
**Pull request by whitevegagabriel**
Added binary that can check for and add Apache 2.0 licenses. Run this binary during the build-rust workflow.

### CODE REVIEW

File: rust/Cargo.toml [hunk 1]
```diff
@@ -24,6 +24,12 @@ itertools = "0.11.0"
 lazy_static = "1.4.0"
 thiserror = "1.0.41"
 
+# Dev tools
+file-header = { version = "0.1.1", optional = true }
+globset = { version = "0.4.13", optional = true }
+# anyhow
+# clap
```
Comment (marshallpierce) [inline rust/Cargo.toml:31]: leftovers?

File: rust/tools/file_header.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+use std::{env, path::PathBuf};
+
+fn main() -> anyhow::Result<()> {
+    let rust_dir = PathBuf::from(env::var("CARGO_MANIFEST_DIR")?);
+    let ignore_globset = ignore_globset()?;
+    // Note: when adding headers, there is a bug where the line spacing is off for Apache 2.0 (see https://github.com/spdx/license-list-XML/issues/2127)
+    let header = APACHE_2_0.build_header(YearCopyrightOwnerValue::new(2023, "Google LLC".into()));
+
+    let cli = Cli::parse();
+
```
Comment (marshallpierce) [inline rust/tools/file_header.rs:30]: rustfmt appears to be upset here

File: rust/tools/file_header.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    APACHE_2_0, YearCopyrightOwnerValue
+}};
+use globset::{Glob, GlobSet, GlobSetBuilder};
+use std::{env, path::PathBuf};
+
+fn main() -> anyhow::Result<()> {
+    let rust_dir = PathBuf::from(env::var("CARGO_MANIFEST_DIR")?);
+    let ignore_globset = ignore_globset()?;
+    // Note: when adding headers, there is a bug where the line spacing is off for Apache 2.0 (see https://github.com/spdx/license-list-XML/issues/2127)
+    let header = APACHE_2_0.build_header(YearCopyrightOwnerValue::new(2023, "Google LLC".into()));
```
Comment (barbibulle) [inline rust/tools/file_header.rs:27]: will that not conflict with headers that have a date range for the copyright? (like `Copyright 2021-2022 Google LLC`)?

Comment (marshallpierce) [reply rust/tools/file_header.rs:27]: The search pattern used to check for an existing license is just the date-free ASL2 header, so it shouldn't care about date ranges: https://github.com/google/file-header/blob/main/src/license/spdx/mod.rs#L327.
However, that does raise the question of perhaps we should be using the current date's year when _adding_ licenses 🤔

File: rust/Cargo.toml (hunk 1 above)
Comment (whitevegagabriel) [reply rust/Cargo.toml:31]: I wanted to signify that these dependencies are used by Dev tools but declared elsewhere. Though I guess its not important. If the dependencies are deleted, the compiler will complain.

File: rust/tools/file_header.rs (hunk 3 above)
Comment (whitevegagabriel) [reply rust/tools/file_header.rs:28]: Ok, so for now we can probably keep as-is.

File: rust/Cargo.toml (hunk 1 above)
Comment (whitevegagabriel) [reply rust/Cargo.toml]: Removed.

**Comment by whitevegagabriel** [comment]
There was an issue in a dependency, so waiting for that to be resolved before I continue here.

File: rust/tools/file_header.rs (hunk 2 above)
Comment (whitevegagabriel) [reply rust/tools/file_header.rs:31]: Should be fixed now

**Comment by whitevegagabriel** [comment]
Ok, issue in dependency has been fixed.

**Comment by barbibulle** [comment]
Ready to mege?

**Comment by whitevegagabriel** [comment]
Ready for merge 👍

**Review by barbibulle — APPROVED** [review]

**Pull request merged**